File: crates/spfs/spfs/cli/_cmd_migrate.py
import os
import argparse

import structlog

import spfs
import spfs.storage.fs.migrations

_logger = structlog.get_logger("cli")


def register(sub_parsers: argparse._SubParsersAction) -> None:

    migrate_cmd = sub_parsers.add_parser("migrate", help=_migrate.__doc__)
    migrate_cmd.add_argument(
        "path", metavar="PATH", help="the path to the filesystem repository to migrate",
    )
    migrate_cmd.set_defaults(func=_migrate)


def _migrate(args: argparse.Namespace) -> None:
    """Migrate the data in and older repository format."""

    repo_root = os.path.abspath(args.path)
    spfs.storage.fs.migrations.migrate_repo(repo_root)

File: crates/spfs-cli/main/src/cmd_migrate.rs
// Copyright (c) Sony Pictures Imageworks, et al.
// SPDX-License-Identifier: Apache-2.0
// https://github.com/imageworks/spk

use anyhow::Result;
use clap::Args;
use spfs::Error;

/// Migrate the data from and older repository format to the latest one
#[derive(Debug, Args)]
pub struct CmdMigrate {
    /// Replace old data with migrated data one complete
    #[clap(long)]
    upgrade: bool,

    /// The path to the filesystem repository to migrate
    path: std::path::PathBuf,
}

impl CmdMigrate {
    pub async fn run(&mut self, _config: &spfs::Config) -> Result<i32> {
        let repo_root = tokio::task::block_in_place(|| dunce::canonicalize(&self.path))
            .map_err(|err| Error::InvalidPath((&self.path).into(), err))?;
        let result = if self.upgrade {
            spfs::storage::fs::migrations::upgrade_repo(repo_root).await?
        } else {
            spfs::storage::fs::migrations::migrate_repo(repo_root).await?
        };
        tracing::info!(path = ?result, "migrated");
        Ok(0)
    }
}
